### PR TITLE
Redesign series block

### PR DESCRIPTION
**Pull request by owi92**
This applies the suggested redesign of the series blocks by adding list and slider view options.
As with the other redesign PR, this is still missing the updated color scheme, which I suggest should be applied once every other part of the redesign is finished.

### CODE REVIEW

File: frontend/src/ui/Blocks/Series.tsx [hunk 1]
```diff
@@ -356,6 +696,9 @@ const GridTile: React.FC<GridTypeProps> = ({ event, basePath, active }) => {
                 {/* `new Date` is well defined for our ISO Date strings */}
                 <RelativeDate date={new Date(date)} isLive={event.isLive} />
             </div>
+            {isList && event.description && <p css={ellipsisOverflowCss(3)}>
+                {event.description}
+            </p>}
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Blocks/Series.tsx:701]: The description should be smaller font and should also say "no description" if there is none. Luckily you can use `SmallDescription` from `src/ui/metadata.tsx`.

File: frontend/src/ui/Blocks/Series.tsx [hunk 2]
```diff
@@ -373,6 +716,7 @@ const GridTile: React.FC<GridTypeProps> = ({ event, basePath, active }) => {
         },
         ...active && {
             backgroundColor: "var(--grey86)",
+            display: "inline-block",
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Blocks/Series.tsx:719]: Can you explain why this change (and the corresponding one below) was necessary?

File: frontend/src/ui/Blocks/Series.tsx [hunk 3]
```diff
@@ -193,46 +210,364 @@ type SeriesBlockContainerProps = {
...
+            }}>
+                {children}
+                {leftVisible && <ProtoButton
+                    onClick={() => scroll(-scrollDistance)}
+                    css={{ position: "absolute", left: 0, ...buttonCss }}>
+                    <FiChevronLeft />
+                </ProtoButton>}
+                {rightVisible && <ProtoButton
+                    onClick={() => scroll(scrollDistance)}
+                    css={{ position: "absolute", right: 0, ...buttonCss }}>
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Blocks/Series.tsx:525]: Also: I would use `right: 8` or sth like that to move the buttons a bit away from the container edge. Looks better IMO.

File: frontend/src/ui/Blocks/Series.tsx [hunk 4]
```diff
@@ -193,46 +210,364 @@ type SeriesBlockContainerProps = {
...
+                    paddingTop: 14,
+                }}>
+                    {title && <div css={{
+                        display: "inline-block",
+                        padding: "8px 12px",
+                    }}>
+                        <WithTooltip tooltip={title}>
+                            <h2 css={{
+                                color: "var(--grey40)",
+                                fontSize: 19,
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Blocks/Series.tsx:244]: I think I prefer this:

```suggestion
                                color: "var(--grey20)",
                                fontSize: 20,
```

File: frontend/src/ui/Blocks/Series.tsx [hunk 5]
```diff
@@ -242,30 +577,33 @@ const UpcomingEventsGrid: React.FC<React.PropsWithChildren> = ({ children }) =>
             backgroundColor: "var(--grey86)",
             borderRadius: 4,
             marginBottom: 8,
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Blocks/Series.tsx:579]: There needs to be more margin above. I like `margin-top: 8`.

image

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]
This looks really good! I think this makes Tobira feel quite a bit nicer. I like almost all of the things you did that were not part of Lisa's design or differ from Lisa's design. The slider view is particularly nice IMO. I totally didn't knew about these scroll-snap thingies. Works really well, also on mobile/with touchscreen. And everything seems to work very nicely with keyboard navigation as well!

However, as usual, I have comments :D Sorry for the large number of them, but all or almost all are very easy to fix, I think. Apart from the line comments, here are some general ones.

- By using tab and enter (keyboard navigation), you can open both dropdowns at the same time.
- The currently active option can still be clicked and is not shown as "active". Maybe we want to change that?
- As you mentioned yourself: the shevron always points down, even if the menu is open. We might still want to fix that, but honestly I don't mind it that much. For what it's worth, in Lisa's design the icon also doesn't change for the open menu.
- When opening one dropdown and then using the WAVE browser plugin, it shows an error.
- In slider view: I can't scroll to the very start. Well I can but then it snaps to first element somehow which is sliiightly farther right than the start. Not a huge deal at all, but yeah.
- I quite like the slider view on mobile -> why not make it possible to use it there? Only the list view makes no sense on mobile I think. Although... even then we could just put the description below the thumbnail and stuff. Actually... I think I'd like that. Then you can get rid of the whole "set state to 'grid' when window gets too small" code. Opinions?
- The left/right buttons on Safari mobile are at the top instead of the center.

And one main one which does not need to be done in this PR: since those choices are completely local, they are reset on navigation and everything. I think that takes a lot of their use from them. In the future we likely want to persist those choices in local storage *somehow* (lotsa open questions) and we also want to give the same three options to the moderator designing the block. So that it can be a good default. But there are still lots of open questions as well. This can be done as an extra PR no problem.

File: frontend/src/i18n/locales/de.yaml [hunk 6]
```diff
@@ -376,7 +384,7 @@ manage:
       series:
         order:
           heading: Reihenfolge
-          new-to-old: Neuste zuerst
+          new-to-old: Neueste zuerst
           old-to-new: Älteste zuerst
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/i18n/locales/de.yaml:388]: Sure thing. I considered using the already existing translation, but that's under `manage.realm.content.series.order.new-to-old` and I wasn't sure if that should be used in this context, so I put a duplicate under `series.settings.new-to-old`. But of course I wouldn't mind using the other one. So just to be clear, would you advise me to delete the longer one and replace it's ocurrance in `frontend/src/routes/manage/Realm/Content/Edit/EditMode/Series.tsx`, or do it the other way around?

File: frontend/src/ui/Blocks/Series.tsx [hunk 7]
```diff
@@ -193,46 +210,364 @@ type SeriesBlockContainerProps = {
...
+const FloatingBaseMenu: React.FC<FloatingBaseMenuProps> = ({
+    triggerContent, list, addTriggerCss, addContainerCss, container,
+}) => (
+    <FloatingContainer
+        ref={container}
+        placement="bottom"
+        trigger="click"
+        ariaRole="menu"
+        distance={0}
+        viewPortMargin={12}
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:282]: Well... no. Truth be told, I initially copied this over from your implementation for the `UserBox` menu :D

File: frontend/src/ui/Blocks/Series.tsx [hunk 8]
```diff
@@ -193,46 +210,364 @@ type SeriesBlockContainerProps = {
...
+    useEffect(() => {
+        setVisibilities();
+    }, []);
+
+    const buttonCss = {
+        alignSelf: "center",
+        backgroundColor: "var(--grey65)",
+        borderRadius: 50,
+        padding: 11,
+        transition: "all .1s ease-in",
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:489]: Nope, no specific reasoning. Just missing some experience and feel for things like this.

File: frontend/src/ui/Blocks/Series.tsx [hunk 9]
```diff
@@ -193,46 +210,364 @@ type SeriesBlockContainerProps = {
...
+    };
+
+    useEffect(() => {
+        setVisibilities();
+    }, []);
+
+    const buttonCss = {
+        alignSelf: "center",
+        backgroundColor: "var(--grey65)",
+        borderRadius: 50,
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:487]: Yeah you're right. I believe I'm just used to using some larger number, as you said, and 50 seemed to work just fine ;)

File: frontend/src/ui/Blocks/Series.tsx (hunk 2 above)
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:719]: Errm it is not. I think I just wondered why this wasn't in here but in the return statement instead, so I played around with it and forgot to change it back.

File: frontend/src/ui/Blocks/Series.tsx [hunk 10]
```diff
@@ -193,43 +210,364 @@ type SeriesBlockContainerProps = {
...
+                marginTop: 24,
+                padding: 12,
+                ...title && { paddingTop: 0 },
+                backgroundColor: "var(--grey95)",
+                borderRadius: 10,
+            }}>
+                <div css={{
+                    display: "flex",
+                    justifyContent: "space-between",
+                    paddingTop: 14,
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:235]: I could give the container of the buttons a `margin-right: 5` or `padding-right: 5` to fix this (the correct value would actually be `4.85` but I think that's basically indistinguishable from `5`).
Or better maybe to just do `padding: 5` and your suggestion of `align-self: flex-start`. This would keep the buttons aligned with a single line title and also make them stick to the top with multi line titles, which I think looks better than having them vertically centered.

File: frontend/src/i18n/locales/de.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply frontend/src/i18n/locales/de.yaml:388]: First option: keep `series.settings.new-to-old`

File: frontend/src/ui/Blocks/Series.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply frontend/src/ui/Blocks/Series.tsx:235]: Yeah I think I like the latter. Works well for single line titles and sticks to the top for multi line ones.

**Comment by LukasKalbertodt** [comment]
Another thing I thought of: on very small screens, the title has very little space (already now, but especially so when using my recommendtations of `whitespace: nowrap` and also showing the "view" setting on small screens). Maybe there should be some special case for that, e.g. set flex direction to `column` for screen smaller than 450 or sth like that?

File: frontend/src/ui/Blocks/Series.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply frontend/src/ui/Blocks/Series.tsx:282]: Ah :D Then maybe remove that line? And maybe also the `distance` line if it still looks good. Using default more is preferable, if it works.

And is the `ariaRole` even correct or does that need to be something else? :thinking: But I guess you have to look into that anyway to fix the WAVE error mentioned in another comment.

[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:282]: I do prefer the `distance={0}` above the default value, which feels a little too much to me without the arrowtip, so I'm keeping this for now. But maybe we want to include the arrowtip after all, even though it differs from the suggested design (which would also require a negative `distance` value, but I don't think we want to have the menu that close to the trigger button).

I believe `ariaRole="menu"` is correct, at least I don't think it should be `listbox`.

File: frontend/src/ui/Blocks/Series.tsx [hunk 11]
```diff
@@ -193,46 +210,364 @@ type SeriesBlockContainerProps = {
...
+    const { t } = useTranslation();
+    const ref = useRef<FloatingHandle>(null);
+    const order = useContext(OrderContext);
+
+    const triggerContent = (
+        <>{order.eventOrder === "NEW_TO_OLD"
+            ? t("series.settings.new-to-old")
+            : t("series.settings.old-to-new")}
+        </>
+    );
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:322]: Hm they do appear to be necessary but I could move them to the prop declaration like this: `triggerContent={<>{triggerContent}</>}`

File: frontend/src/ui/Blocks/Series.tsx [hunk 12]
```diff
@@ -193,46 +210,364 @@ type SeriesBlockContainerProps = {
...
+        gallery: () => <FiGrid />,
+        list: () => <FiList />,
+    });
+
+    const triggerContent = (
+        <div css={{
+            display: "flex",
+            alignItems: "center",
+            svg: { fontSize: 22, color: "var(--grey40)" },
+        }}>{icon}</div>
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:354]: Giving the parent button element an `aria-label` apparently also works.

File: frontend/src/ui/Blocks/Series.tsx [hunk 13]
```diff
@@ -193,46 +210,364 @@ type SeriesBlockContainerProps = {
...
+            outline: "2px solid var(--accent-color)",
+        },
+    };
+
+    return (
+        <div css={{ position: "relative" }}>
+            <div tabIndex={-1} onScroll={() => setVisibilities()} ref={ref} css={{
+                display: "grid",
+                gridAutoFlow: "column",
+                gridAutoColumns: 230,
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:508]: True. I saw this being done somewhere else with grid, and wanted to try it, but you're right, this should be flex. That also seems to fix the button position for safari mobile views.

File: frontend/src/ui/Blocks/Series.tsx [hunk 14]
```diff
@@ -193,46 +210,364 @@ type SeriesBlockContainerProps = {
...
+            if (window.innerWidth < VIDEO_GRID_BREAKPOINT) {
+                setViewState("gallery");
+            }
+        };
+        window.addEventListener("resize", handleResize);
+
+        return () => {
+            window.removeEventListener("resize", handleResize);
+        };
+    });
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:553]: Yeah, I opted to remove this instead, since you suggested it in your top comment.

**Comment by owi92** [comment]
Ok, so I _believe_ I managed to adress all your comments, but take that with a grain of salt. Unfortunately I forgot to do this a little more atomically, so it might again be a little frustrating to review - sorry for that! And if that is a larger issue, I could of course go back and make smaller commits out of these changes. Also, when pushing the rebased status a little earlier today, I forgot that I already did an amend on the first commit of this - but that is only the reversion of the breadcrumb changes.

Note that this is still missing the ability to use the menus with arrow keys.

File: frontend/src/i18n/locales/en.yaml [hunk 15]
```diff
@@ -163,6 +163,18 @@ series:
     text: >
       The data of the requested series has not been fully transferred yet. This should
       happen automatically soon. Try again in a few minutes.
+  settings:
+    order: Order
+    order-label: Choose video order
+    old-to-new: Oldest first
```
Comment (LukasKalbertodt) [inline frontend/src/i18n/locales/en.yaml:169]: `new-to-old` mysteriously disappeared (this is also obvious when checking test deployment in English)

File: frontend/src/ui/Blocks/Series.tsx [hunk 16]
```diff
@@ -193,46 +218,412 @@ type SeriesBlockContainerProps = {
...
+                <div css={{
+                    display: "flex",
+                    [`@media (max-width: ${VIDEO_GRID_BREAKPOINT}px)`]: {
+                        flexDirection: "column",
+                    },
+                }}>
+                    {title && <div css={{
+                        display: "inline-block",
+                        padding: "8px 12px",
+                    }}>
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Blocks/Series.tsx:253]: Maybe in one line? Or... even better: can we remove this div and just add the CSS to the `h2`?

File: frontend/src/ui/Blocks/Series.tsx [hunk 17]
```diff
@@ -193,46 +218,412 @@ type SeriesBlockContainerProps = {
...
+                        padding: 5,
+                        [`@media not all and (max-width: ${VIDEO_GRID_BREAKPOINT}px)`]: {
+                            marginLeft: "auto",
+                        },
+                        [`@media (max-width: ${VIDEO_GRID_BREAKPOINT}px)`]: {
+                            width: "100%",
+                            justifyContent: "space-between",
+                            paddingLeft: 11,
+                            paddingRight: 11,
+                        },
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Blocks/Series.tsx:275]: Is it intended that the two buttons have space between them? I would think it looks better if they are both on the right.

So I would remove all of these lines. And add `justifyContent: flex-end` but without media query. The replace the `margin-left` here, the parent div should have `justify-content: space-between`. Unless I'm missing something that works better with less CSS and it looks better IMO.

The only "problem" this is that if the title is missing, the buttons are on the left again. I think I would solve that by just removing `title &&`. So that the `<div><h2 /></div>` (or even just `<h2 />` as mentioned above) is always rendered, but possibly with nothing inside.

File: frontend/src/ui/Blocks/Series.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply frontend/src/ui/Blocks/Series.tsx]: We can try the arrow tips :shrug: I think both can work. And yeah I agree the distance in the design is too low, and we dont want that.

File: frontend/src/ui/Blocks/Series.tsx [hunk 18]
```diff
@@ -394,6 +792,14 @@ const GridTile: React.FC<GridTypeProps> = ({ event, basePath, active }) => {
                 boxShadow: "0 0 0 2px var(--accent-color)",
             },
         },
+        // ListView styles:
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Blocks/Series.tsx:795]: List view on mobile is broken still: the description and all that stuff needs to go under the thumbnail on small screens.

File: frontend/src/ui/Blocks/Series.tsx (hunk 17 above)
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:275]: Yeah I did `space-between` because I found the situation depicted below to look a little awkward when the title is short, but I suppose it's not that bad. I could also just use a lower breakpoint for `flex-direction: column`, or align (or rather "justify") the buttons with the heading on the left on smaller screens (second screenshot). Let me know what you prefer. 

Bildschirm­foto 2023-02-20 um 15 25 29


Bildschirm­foto 2023-02-20 um 15 32 22

I believe that `justify-content: flex-end` does not work with `align-self: flex-start` (which I use to align the buttons vertically with the first line of multiline headings), but we can use `margin-left: auto` instead of `justify-content` to keep the buttons on the right.

Comment (LukasKalbertodt) [reply frontend/src/ui/Blocks/Series.tsx:275]: Ah mh, thats fair, it does look weird for small titles :/ 
What about using `flex-wrap: wrap` instead of `flex-direction: column`? Then the second line is only created when there is actually not enough space. Specific suggestion:

- Remove all 5 CSS properties that this review comment is attached to.
- Add `margin-left: auto` to the button container (for all screen sizes)
- For width < breakpoint, add `flex-wrap: wrap` for the outer container (and remove `flex-direction: column`)

For any given title, there is still a screen width where it looks weird (since it juuust wrapped around), but thats better than the previous idea. What do you think?

File: frontend/src/ui/Blocks/Series.tsx [hunk 19]
```diff
@@ -193,46 +218,412 @@ type SeriesBlockContainerProps = {
...
+};
+
+const MenuItem: React.FC<MenuItemProps> = ({ Icon, label, onClick, close, disabled }) => {
+    const ref = useRef<HTMLButtonElement>(null);
+    const preventEnter = (event: React.KeyboardEvent<HTMLButtonElement>) => {
+        if (ref.current?.getAttribute("aria-disabled") === "true"
+                && event.key === "Enter") {
+            event.preventDefault();
+        }
+    };
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:488]: Then we could also just use `disabled` on the buttons instead of `aria-disabled`, right?
The reasoning behind this (admittedly not too pretty) function was to specifically keep the ability to focus the disabled buttons as I was under the impression that this would be the right approach for accessibility concerns. But now that I'm trying to find a good reference, I struggle to do so and realize that I might have misunderstood or misremembered some things about this.

File: frontend/src/ui/Blocks/Series.tsx (hunk 17 above)
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:275]: Sure, that works for me!

**Comment by owi92** [comment]
One last thing that still bugs me are the different margins between the `block-header` and the `upcoming events grid` (and to a lesser extent also the left/right margin difference of the header). We could adjust these but then the videos from the `upcoming events` wouldn't align with the videos from the regular grid unless the upcoming events had less padding. 😖 But I guess that doesn't need to be done or even decided in this PR.
Bildschirm­foto 2023-02-17 um 16 59 10

File: frontend/src/ui/Blocks/Series.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply frontend/src/ui/Blocks/Series.tsx]: Ah interesting. Yeah I think my intuition would be to not make it tab-selectable if it's disabled. But I also have no source on whether that's the right thing to do. Maybe let's go with the simpler option for now.

File: frontend/src/ui/Blocks/Series.tsx (hunk 17 above)
Comment (LukasKalbertodt) [reply frontend/src/ui/Blocks/Series.tsx]: I like how it works now I think. The `justifyContent: "space-between"` on the parent div is not even required anymore.

File: frontend/src/ui/Blocks/Series.tsx [hunk 20]
```diff
@@ -193,46 +218,394 @@ type SeriesBlockContainerProps = {
...
+    const ref = useRef<HTMLButtonElement>(null);
+
+    return <li css={{
+        ":not(:last-child)": {
+            borderBottom: "1px solid var(--grey86)",
+        },
+        ":last-child button": {
+            borderRadius: "0 0 8px 8px",
+        },
+    }}>
```
[the remark this answers is not in the document]
Comment (owi92) [reply frontend/src/ui/Blocks/Series.tsx:473]: Oh no, that's on me. I first removed the parentheses and after that moved the child styles down there without checking the indentations again.

**Comment by owi92** [comment]
Ok, so there is a simple way to check if the children of the `SeriesBlockContainer` contain any actual events, but it feels kinda hacky... I included it for now, but I'm thinking there must be a nicer way that I am missing.

Regarding the arrow tips: I included them in the previous iteration because I was curious about your opinion. But I also prefer the menus without them, so I removed them again.

**Comment by LukasKalbertodt** [comment]
- Yeah, it's nicer without arrow tips.
- And I agree, the way you check whether there are events is a bit too hacky for my taste. I would just add a `hasEvents: boolean` prop to the component which can be passed down directly from the parent as that parent has access to the events array.
- Regarding the re-indentation you did: in all those case both options are totally fine to me. So yeah, just keep it as it is now.
- One more thing I noticed: can you add `lines={3}` to the `<SmallDescription>`? The default is 2, but 3 fit no problem in the desktop view. So we should show that additional information.

File: frontend/src/ui/Blocks/Series.tsx [hunk 21]
```diff
@@ -132,6 +157,8 @@ const ReadySeriesBlock: React.FC<ReadyProps> = ({
     const upcomingLiveEvents = series.events.filter(event =>
         isUpcomingLiveEvent(event.syncedData?.startTime ?? null, event.isLive));
 
+    const eventsNotEmpty = events.length > 0 || upcomingLiveEvents.length > 0;
```
Comment (owi92) [inline frontend/src/ui/Blocks/Series.tsx:160]: Hm this could also just check for `series.events.length > 0`

File: frontend/src/ui/Blocks/Series.tsx [hunk 22]
```diff
@@ -162,7 +189,7 @@ const ReadySeriesBlock: React.FC<ReadyProps> = ({
             {...{ basePath, event }}
         />);
 
-    const eventsUI = events.length === 0
+    const eventsUI = !eventsNotEmpty
```
Comment (owi92) [inline frontend/src/ui/Blocks/Series.tsx:192]: Added this check because before we would show the `no-events` text even if there were upcoming events present

**Review by LukasKalbertodt — APPROVED** [review]
Wonderful! :tada:
